File: src/test/java/com/esentri/wrabbitmq/java/SendListenTest.java
package com.esentri.wrabbitmq.java;

import com.esentri.wrabbitmq.WrabbitTopic;
import org.junit.jupiter.api.Test;

import java.io.Serializable;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

import static com.esentri.wrabbitmq.java.TestFunctions.Await;
import static org.fest.assertions.Assertions.assertThat;

public class SendListenTest {

   private final WrabbitTopic topic = new WrabbitTopic("TestTopic-Listener-Java");

   private <MESSAGE extends Serializable> WrabbitEvent<MESSAGE> newEvent() {
      return new WrabbitEvent<>(topic, UUID.randomUUID().toString());
   }

   @Test
   public void sendListenString() {
      final WrabbitEvent<String> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(1);
      final String message = "1: Hello World!";

      event.listener(it -> {
         assertThat(it).isEqualTo(message);
         countDownLatch.countDown();
      });
      event.send(message);

      Await(countDownLatch);
   }

   @Test
   public void sendListenString_X_times() {
      final WrabbitEvent<String> event = newEvent();
      final String message = "2: Hello World!";
      final int sentTimes = 1000;
      final CountDownLatch countDownLatch = new CountDownLatch(sentTimes);

      event.listener(it -> {
         assertThat(it).isEqualTo(message);
         countDownLatch.countDown();
      });
      IntStream.range(0, sentTimes).forEach(it ->
         event.send(message)
      );

      Await(countDownLatch);
   }

   @Test
   public void sendListenString_2_listener() {
      final WrabbitEvent<String> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(2);
      final String message = "3: Hello World!";

      event.listener(it -> {
         assertThat(it).isEqualTo(message);
         countDownLatch.countDown();
      });
      event.listener(it -> {
         assertThat(it).isEqualTo(message);
         countDownLatch.countDown();
      });
      event.send(message);

      Await(countDownLatch);
   }

   @Test
   public void sendTestObjectObject() {
      final WrabbitEvent<TestObjectObject> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(1);
      final TestObjectObject message = new TestObjectObject(new TestObjectNumberText(12345, "1: Hello World!"));

      event.listener(it -> {
         assertThat(it).isInstanceOf(TestObjectObject.class);
         assertThat(it.getObj()).isInstanceOf(TestObjectNumberText.class);
         assertThat(it.getObj().getNumber()).isEqualTo(message.getObj().getNumber());
         assertThat(it.getObj().getText()).isEqualTo(message.getObj().getText());
         countDownLatch.countDown();
      });
      event.send(message);

      Await(countDownLatch);
   }

   @Test
   public void sendParallel() {
      final WrabbitEvent<String> event1 = newEvent();
      final WrabbitEvent<TestObjectObject> event2 = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(2);
      final String message1 = "3: Hello World!";
      final TestObjectObject message2 = new TestObjectObject(new TestObjectNumberText(12345, "2: Hello World!"));

      event1.listener(it -> {
         assertThat(it).isEqualTo(message1);
         countDownLatch.countDown();
      });
      event2.listener(it -> {
         assertThat(it).isInstanceOf(TestObjectObject.class);
         assertThat(it.getObj()).isInstanceOf(TestObjectNumberText.class);
         assertThat(it.getObj().getNumber()).isEqualTo(message2.getObj().getNumber());
         assertThat(it.getObj().getText()).isEqualTo(message2.getObj().getText());
         countDownLatch.countDown();
      });
      event1.send(message1);
      event2.send(message2);

      Await(countDownLatch);
   }

   @Test
   public void sendGroups() throws InterruptedException {
      final WrabbitEvent<String> event = newEvent();
      // AtomicInteger is used to check if after the expected events others will occure.
      // This would not be possible with CountDownLatch.
      final AtomicInteger waitCounter = new AtomicInteger(0);
      final String message = "5: Hello World!";
      final int sentTimes = 10;

      event.listener("group1", it -> {
         assertThat(it).isEqualTo(message);
         waitCounter.incrementAndGet();
      });
      event.listener("group1", it -> {
         assertThat(it).isEqualTo(message);
         waitCounter.incrementAndGet();
      });
      event.listener("group2", it -> {
         assertThat(it).isEqualTo(message);
         waitCounter.incrementAndGet();
      });

      IntStream.range(0, sentTimes).forEach(it -> event.send(message));

      while(waitCounter.get() < sentTimes * 2) {
         Thread.sleep(300);
      }

      for(int i = 0; i < 5; i++) {
         Thread.sleep(500);
      }

      assertThat(waitCounter.get()).isEqualTo(sentTimes * 2);
   }

   @Test
   public void sendAndListenWithContext() {
      final WrabbitEvent<String> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(1);
      final String message = "6: Hello World!";
      String propertyKey = "test";
      String propertyValue = "property";

      event.listener((context, it) -> {
         assertThat(it).isEqualTo(message);
         assertThat(context.get(propertyKey).toString()).isEqualToIgnoringCase(propertyValue);
         countDownLatch.countDown();
      });
      event
         .messageBuilder()
         .property(propertyKey, propertyValue)
         .send(message);

      Await(countDownLatch);
   }
}

File: src/test/kotlin/com/esentri/wrabbitmq/SendListenTest.kt
package com.esentri.wrabbitmq

import org.fest.assertions.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.io.Serializable
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicInteger

class SendListenTest {

   private val topic = WrabbitTopic("TestTopic-Listeners")

   private fun <MESSAGE: Serializable> newEvent(): WrabbitEvent<MESSAGE> =
     WrabbitEvent(topic, UUID.randomUUID().toString())

   @Test
   fun sendListenString() {
      val wrabbitEvent = newEvent<String>()
      val countDownLatch = CountDownLatch(1)
      val message = "1: Hello World!"

      wrabbitEvent.listener { it ->
         assertThat(it).isEqualTo(message)
         countDownLatch.countDown()
      }
      wrabbitEvent.send(message)

      Await(countDownLatch)
   }

   @Test
   fun sendListenString_X_times() {
      val wrabbitEvent = newEvent<String>()
      val sentTimes = 1000
      val countDownLatch = CountDownLatch(sentTimes)
      val message = "2: Hello World!"

      wrabbitEvent.listener { it ->
         assertThat(it).isEqualTo(message)
         countDownLatch.countDown()
      }
      for (i in 1..sentTimes) {
         wrabbitEvent.send(message)
      }

      Await(countDownLatch)
   }

   @Test
   fun sendListenString_2_listener() {
      val wrabbitEvent = newEvent<String>()
      val countDownLatch = CountDownLatch(2)
      val message = "3: Hello World!"

      wrabbitEvent.listener { it ->
         assertThat(it).isEqualTo(message)
         countDownLatch.countDown()
      }
      wrabbitEvent.listener { it ->
         assertThat(it).isEqualTo(message)
         countDownLatch.countDown()
      }
      wrabbitEvent.send(message)

      Await(countDownLatch)
   }

   @Test
   fun sendTestObjectObject() {
      val wrabbitEvent = newEvent<TestObjectObject>()
      val countDownLatch = CountDownLatch(1)
      val message = TestObjectObject(TestObjectNumberText(12345, "1: Hello World!"))

      wrabbitEvent.listener { it ->
         assertThat(it).isInstanceOf(TestObjectObject::class.java)
         assertThat(it.obj).isInstanceOf(TestObjectNumberText::class.java)
         assertThat(it.obj.number).isEqualTo(message.obj.number)
         assertThat(it.obj.text).isEqualTo(message.obj.text)
         countDownLatch.countDown()
      }
      wrabbitEvent.send(message)

      Await(countDownLatch)
   }

   @Test
   fun sendParallel() {
      val wrabbitEvent1 = newEvent<String>()
      val wrabbitEvent2 = newEvent<TestObjectObject>()
      val countDownLatch = CountDownLatch(2)
      val message1 = "4: Hello World!"
      val message2 = TestObjectObject(TestObjectNumberText(12345, "2: Hello World!"))

      wrabbitEvent1.listener { it ->
         assertThat(it).isEqualTo(message1)
         countDownLatch.countDown()
      }
      wrabbitEvent2.listener { it ->
         assertThat(it).isInstanceOf(TestObjectObject::class.java)
         assertThat(it.obj).isInstanceOf(TestObjectNumberText::class.java)
         assertThat(it.obj.number).isEqualTo(message2.obj.number)
         assertThat(it.obj.text).isEqualTo(message2.obj.text)
         countDownLatch.countDown()
      }

      wrabbitEvent1.send(message1)
      wrabbitEvent2.send(message2)

      Await(countDownLatch)
   }

   @Test
   fun sendGroups() {
      val wrabbitEvent = newEvent<String>()
      // AtomicInteger is used to check if after all expected events another one is arriving
      // This would not be possible with CountDownLatch
      val waitCounter = AtomicInteger(0)
      val sentTimes = 10
      val message = "5: Hello World!"

      wrabbitEvent.listener("group1") { it ->
         assertThat(it).isEqualTo(message)
         waitCounter.incrementAndGet()
      }
      wrabbitEvent.listener("group1") { it ->
         assertThat(it).isEqualTo(message)
         waitCounter.incrementAndGet()
      }
      wrabbitEvent.listener("group2") { it ->
         assertThat(it).isEqualTo(message)
         waitCounter.incrementAndGet()
      }

      for (i in 1..sentTimes) {
         wrabbitEvent.send(message)
      }
      while (waitCounter.get() < sentTimes * 2) {
         Thread.sleep(300)
      }
      for (i in 1..5) {
         Thread.sleep(500)
      }
      assertThat(waitCounter.get()).isEqualTo(sentTimes * 2)
   }

   @Test
   fun sendAndListenWithContext() {
      val wrabbitEvent = newEvent<String>()
      val countDownLatch = CountDownLatch(1)
      val message = "6: Hello World!"
      val propertyKey = "test"
      val propertyValue = "property"

      wrabbitEvent.listener { context, it ->
         assertThat(it).isEqualTo(message)
         assertThat(context[propertyKey].toString()).isEqualToIgnoringCase(propertyValue)
         countDownLatch.countDown()
      }
      wrabbitEvent
         .messageBuilder()
         .property(propertyKey, propertyValue)
         .send(message)

      Await(countDownLatch)
   }
}